### DISCUSSION TITLE
Multi Tenancy Setup supporting connections to multiple DB instances simultaneously

### DISCUSSION THREAD

**Discussion by sbuss1980** [prisma/prisma#20920] [Q&A] [answered]
### Question

As recommended by @janpio : moved / copied from: https://github.com/prisma/prisma/issues/2443#issuecomment-1698702617

We are currently evaluating prisma and are discussing possible architectures for a solution.

The current approach is based von Next.JS, Prisma and MSSQL Server and involves the need that users can connect to different databases by configuration at runtime. So a user might have access to multiple databases by permission provided from a different backend and can choose at runtime to which db she wants to connect to. For now I'd assume that schema is structure-wise identical for all. So we wouldn't need to generate and maintain multiple prisma schemas - at least not for this. Potential user base is some 100s up to 1,000 of users, for concurrent users I wouldn't assume more than 50 (rather less). 

So we are wondering if there is a way to implement this using prisma.
From reading the docs and following discussions here it appears technically possible, but we are wondering if it is a reasonable approach *non-functionally, performance-wise* to 
a) First Approach: alter the prisma client instance by request and user everytime it is requested (resp. when a different db is to be queried based on user's settings). This doesn't appear ideal performance-wise assuming that there is one shared prisma client instance for all users / sessions, which would have to be reinitialised every time. Or am I wrong about this?
b) Second Approach: maintain a map or comparable structure of prisma client instances in the nextjs app / node server instance in memory, one for each db / schema instance requested? That can obviously result in numerous prisma client instances based on the number of dbs/schemas requested. And would infer some complexity because we'd have to take care of purging the list when clients are not used for a certain amout of time. etc. The risks which come to mind are of course memory issues when there are too many prisma client instances etc., and also connection pooling issues on db side

Is there any experience towards any of the approaches here? 

Obvious alternatives would be:
* merge all data into one db schema instance (somewhat difficult for some reasons lying in the nature of the app / use cases)
* Scale horizontally - one node instance per DB schema connection (appears very time consuming and expensive, no idea how to implement in a way that it creates an acceptable ux).

Any advice, recommendations are appreciated! Thank you!

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by janpio** [comment, marked as answer]
You definitely want to go with b) here. 

Creating a database connection is costly, and you can not easily switch between databases on runtime. So I would suggest you keep a list of Prisma Clients and make sure to only use the correct one for the correct user. Prisma Client is not optimized for this use case, so it will use a bunch of memory for each instance - so you might need to manage the clients you have open. Usually that goes fine though and turns out not to be that big of a problem.

https://github.com/prisma/prisma/issues/12420 is a feature request to build something specifically for such use cases that uses database _schemas_ instead of databases. Prisma can switch between those dynamically, without closing and re-opening connections, so that could be faster - but we have not implemented that yet.
And https://github.com/prisma/prisma/issues/2077 is a generic issue for this use case where each "tenant" should have its own namespace with data, database or schema.

**Reply by sbuss1980** [reply]
Thanks @janpio !
I have subscribed to both #12420 and #2077. Having taken a look at https://prismaio.notion.site/Prisma-ORM-Roadmap-50766227b779464ab98899accb98295f none of them appear to be available soon. So we'll discuss and most probably give b) a try then, especially if you say, that this shouldn't cause too much trouble resource-wise.

**Reply by ACronje** [reply]
@sbuss1980 I'm curious to hear how this has worked out for you as I have the same problem to solve. Thanks!

**Reply by artemchs** [reply]
Would there be any kind of tutorial or explanation in the documentation for the preferred way to do so?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
